registry/confdb_control.go

```diff
@@ -0,0 +1,141 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package registry
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+)
+
+var (
+	validAccountID = regexp.MustCompile("^(?:[a-z0-9A-Z]{32}|[-a-z0-9]{2,28})$")
+)
+
+type AuthenticationMethod string
+
+const (
+	OperatorKey AuthenticationMethod = "operator-key"
+	Store       AuthenticationMethod = "store"
+)
+
+// isValidAuthenticationMethod checks if a string is a valid AuthenticationMethod
+func isValidAuthenticationMethod(value string) bool {
+	switch AuthenticationMethod(value) {
+	case OperatorKey, Store:
+		return true
+	default:
+		return false
+	}
+}
+
+// convertToAuthenticationMethod converts and validates a []string to []AuthenticationMethod
+func convertToAuthenticationMethod(methods []string) ([]AuthenticationMethod, error) {
+	sort.Slice(methods, func(i, j int) bool {
+		return methods[i] < methods[j]
+	})
+
+	// remove duplicates
+	methods = compact(methods)
+
+	var result []AuthenticationMethod
+	for _, method := range methods {
+		if !isValidAuthenticationMethod(method) {
+			return nil, fmt.Errorf("invalid authentication method: %s", method)
+		}
+		result = append(result, AuthenticationMethod(method))
+	}
+	return result, nil
+}
+
+// Operator holds the delegations for a single operator.
+type Operator struct {
+	ID     string
+	Groups []*ControlGroup
+}
+
+// ControlGroup holds a set of views delegated through the given authentication.
+type ControlGroup struct {
+	Authentication []AuthenticationMethod
+	Views          []string
```

Comment:
shouldn't this contain structs already parsed into  account/confdb/view ?

Reply:
In the assertion, the list of groups is flat.. like so:

```yaml
groups:
  - operator-id:        john
    authentication: 	[ operator-key, store ]
    views:
      - a/b/c
      - d/e/f
  - operator-id:        john
    authentication: 	[ store ]
    views:
      - x/y/z
```

Internally though, we're grouping by operator to make this easier to look up later since all the API operations revolve around an operator. For instance, `isDelegated(operator, view)`, `delegate(operator, views, authMethods)`, `revoke(operator, ...)`.

```yaml
operators:
  john:
    ID: john
    Groups:
      - Authentication: [ operator-key, store ]
        Views:
          - a/b/c
          - d/e/f
      - Authentication: [ store ]
        Views:
          - x/y/z
```

Edit:

Or do you mean we should store the `Views` as a "set" for faster lookup.. Something like `map[string]struct{}`?

Reply:
I mean of doing something like this after the parsing:

```
Views []*ViewRef

type ViewRef struct {
   Account string
   Confdb string
   View string
}
```
or similar

does this make sense to you @miguelpires as well ?

Reply:
Ah, okay. That makes sense. I've made the necessary changes to reflect this.